### ISSUE TITLE
[Theory] Anchors

### ISSUE THREAD

**Issue by adeyris** [fizyr/keras-retinanet#323]
Hi ! Thanks for this network.
I've been reading the paper, and others to understand how the network is built.
I understand the architecture, but i can't get how the predictions are made.
After the pyramid is built, we create the anchors on each of its layers, and during training, these anchors are compared to gt boxes. If their IoU > 0.5 , then they are assigned to the label corresponding to the gt box, if their Iou < 0.04 they are assigned to background. Then these anchor boxes are fed through the classification network, and the output is compared to the label they've been assigned to in the previous step, then backpropagation etc..
Until here am I right ?
Then i can't understand how the regression submodel works, could you please try to explain ?

Thank you, have a good weekend !

**Comment by hgaiser** [comment]
> Until here am I right ?

Yep, you're right.

> Then i can't understand how the regression submodel works, could you please try to explain ?

The regression submodel works similarly. Each anchor is a box in image space. If an anchor is assigned to an annotation (as you said, using the IoU > 0.5 method), the difference between the anchor box and the annotation box is computed. This is encoded as a delta for the center and a delta for width and height, so 4 values in total. There's a bit of extra magic with a log or something like that, but in essence it is the delta for these values.

The regression submodel generates these delta's as well. The loss of the network is the difference between the regression targets and the predicted regression values.

**Comment by adeyris** [comment]
Hi ! 

Thanks for you answer.
What i can't understand is how can the regression model predict ? What's the input of the network ? Is it only the anchor, if it is, how can the network predict without seeing what's around the anchor ?

Thank you very much for your time

**Comment by adeyris** [comment]
Ok, and do these features have anything to do with the anchors ?

**Comment by adeyris** [comment]
Ok, I thought the input was only an anchor. So the input layer is connected to the feature map computed by the FPN, and for each pixel of this feature map, it computes the delta x,y,w,h offset to apply ?

**Comment by adeyris** [comment]
More precisely, I don't understand if one pass through the network predicts one bounding box for one anchor, or if one pass predicts the coordinates of all the anchors in one feature map ?

**Comment by hgaiser** [comment]
The latter. One pass predicts coordinates and classes for all anchors.

I'll close this since it's not an issue and it appears to be answered. For further questions about how retinanet works, please join the keras-retinanet slack channel.
